gtdynamics/pandarobot/ikfast/ikfastwrapper.h

```diff
@@ -0,0 +1,188 @@
+     * 
+     * @param joint_config 
+     * @return std::vector<IkReal> 
+     */
+    ikvector IkFastWrapper::forward(ikvector joint_config) {
+
+        // Probably should be an assert
+        if( joint_config.size() != num_joints ) {
+            std::cout << "\nError: (forward IkFastWrapper) expects vector of "<< num_joints << " values describing joint angles (in radians).\n\n";
+            return ikvector(0);
+        }
+
+        // This will be changed once completely integrated
+        // Put input joint values into array, could be better
+        IkReal joints[num_joints];
+        for (int i=0; i<num_joints; i++)
+        {
+            joints[i] = joint_config[i];
+        }
+
+        IkReal eerot[9],eetrans[3];
+        IKFAST_NAMESPACE::ComputeFk(joints, eetrans, eerot);
+
+
+        // This will be changed once completely integrated
+        // Join rotation matrix and position vector into pose vector
+        ikvector ee_pose(12);
+        for (int i=0; i<3; i++) {
+            for (int j=0; j<3; j++)
+                ee_pose[4*i+j] = eerot[i*3+j];
+            ee_pose[4*i+3] = eetrans[i];
+        }
+
+        return ee_pose;
+    }
+    
+    /**
+     * @brief Inverse IkFastWrapper on Panda robot using IKFast.
+     * 
+     * @param ee_pose 
+     * @param vfree 
+     * @return std::vector<std::vector<IkReal>>
+     */
+    std::vector<ikvector> IkFastWrapper::inverse(ikvector ee_pose, ikvector vfree) {
+        
+
+        // This will be changed once completely integrated
+        // Probably should be an assert
+        if (!(ee_pose.size() == 12 && vfree.size()==num_free_params)){
+                    std::cout << "\nError: (inverse IkFastWrapper) please specify transformation of end effector with one of the following formats:\n"
+                  << "    1) A vector of 7 values: a 3x1 translation (tX), and a 1x4 quaternion (w + i + j + k)\n"
+                  <<  "    2) A (row-major) vector of 12 values: a 3x4 rigid transformation matrix with a 3x3 rotation R (rXX), and a 3x1 translation (tX)\n\n" ;
+            return std::vector<ikvector>();
+        }
+
+        // This will be changed once completely integrated
+        // Separate pose into arrays: rotation matrix and position vector
+        IkReal eerot[9],eetrans[3];
+        for (unsigned int i=0; i<3; i++) {
+            for (int j=0; j<3; j++)
+                eerot[i*3+j] = ee_pose[4*i+j];
+            eetrans[i] = ee_pose[4*i+3];
+        }
+
+        // Ikfast class where solutions are stored
+        IkSolutionList<IkReal> solutions;
+
+        bool bSuccess = IKFAST_NAMESPACE::ComputeIk(eetrans, eerot, vfree.size() > 0 ? &vfree[0] : NULL, solutions);
+
+        if( !bSuccess ) {
+            fprintf(stderr,"Error: (inverse IkFastWrapper) failed to get ik solution\n");
+            return std::vector<ikvector>();
+        }
+
+        int num_sols = (int)solutions.GetNumSolutions();
+
+        // This will be changed once completely integrated
```

Comment:
Just return the vector of Vector7, empty if there is only singularities or outside workspace